src/bp/nlu-server/api.ts

```diff
@@ -0,0 +1,237 @@
+import bodyParser from 'body-parser'
+import { NLU } from 'botpress/sdk'
+import cors from 'cors'
+import express, { Application } from 'express'
+import rateLimit from 'express-rate-limit'
+import { createServer } from 'http'
+import { validate } from 'joi'
+import _ from 'lodash'
+import ms from 'ms'
+import Engine from 'nlu-core/engine'
+
+import { NLUServerLogger } from './logger'
+import ModelService from './model-service'
+import { monitoringMiddleware, startMonitoring } from './monitoring'
+import TrainService from './train-service'
+import TrainSessionService from './train-session-service'
+import { TrainInput } from './typings'
+import { authMiddleware, handleErrorLogging, handleUnexpectedError } from './util'
+import { TrainInputCreateSchema } from './validation'
+
+export type APIOptions = {
+  version: string
+  host: string
+  port: number
+  modelDir: string
+  authToken?: string
+  limitWindow: string
+  limit: number
+}
+
+const debug = DEBUG('api')
+const debugRequest = debug.sub('request')
+
+const createExpressApp = (options: APIOptions): Application => {
+  const app = express()
+
+  // This must be first, otherwise the /info endpoint can't be called when token is used
+  app.use(cors())
+
+  app.use(bodyParser.json({ limit: '250kb' }))
+
+  app.use((req, res, next) => {
+    res.header('X-Powered-By', 'Botpress')
+    debugRequest('incoming ' + req.path, { ip: req.ip })
+    next()
+  })
+
+  app.use(monitoringMiddleware)
+  app.use(handleUnexpectedError)
+
+  if (process.core_env.REVERSE_PROXY) {
+    app.set('trust proxy', process.core_env.REVERSE_PROXY)
+  }
+
+  if (options.limit > 0) {
+    app.use(
+      rateLimit({
+        windowMs: ms(options.limitWindow),
+        max: options.limit,
+        message: 'Too many requests, please slow down'
+      })
+    )
+  }
+
+  if (options.authToken?.length) {
+    app.use(authMiddleware(options.authToken))
+  }
+
+  return app
+}
+
+export default async function(options: APIOptions) {
+  const app = createExpressApp(options)
+  const logger = new NLUServerLogger('API')
+  const loggerWrapper: NLU.Logger = {
+    info: (msg: string) => logger.info(msg),
+    warning: (msg: string, err?: Error) => (err ? logger.attachError(err).warn(msg) : logger.warn(msg)),
+    error: (msg: string, err?: Error) => (err ? logger.attachError(err).error(msg) : logger.error(msg))
+  }
+
+  let engine: Engine
+  let modelService: ModelService
+  let trainSessionService: TrainSessionService
+  let trainService: TrainService
+  try {
+    engine = new Engine('nlu-server', loggerWrapper)
+    modelService = new ModelService(options.modelDir)
+    await modelService.init()
+    trainSessionService = new TrainSessionService()
+    trainService = new TrainService(logger, engine, modelService, trainSessionService)
```

Comment:
Constructors shouldn't be able to fail, please move them out of the try/catch block.